src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -373,29 +372,35 @@ ExpressionSyntax initializer
 			TypeSyntax TypeSyntax,
 			SyntaxToken Identifier,
 			ExpressionSyntax Initializer,
-			bool IsAutoImplemented,
-			bool IsReadOnly
-		) GetPropertyStuff( IPropertySymbol symbol, SyntaxNode syntax )
+			bool IsAutoImplemented
+		) GetPropertyStuff( SyntaxNode syntax )
 			=> syntax switch {
 				PropertyDeclarationSyntax prop => (
 					prop.Type,
 					prop.Identifier,
 					prop.Initializer?.Value,
-					prop.IsAutoImplemented(),
-					symbol.IsReadOnly
+					prop.IsAutoImplemented()
 				),
 				ParameterSyntax param => (
 					param.Type,
 					param.Identifier,
 					Initializer: null,
-					IsAutoImplemented: true,
-
-					// The property X in "record Foo(int X);" is not IsReadOnly,
-					// but its SetMethod has IsInitOnly. We could probably just
-					// return true here but this seems safer.
-					IsReadOnly: symbol.SetMethod != null && symbol.SetMethod.IsInitOnly
+					IsAutoImplemented: true
 				),
 				_ => throw new NotImplementedException()
 			};
+
+		private bool IsReadOnly( IPropertySymbol symbol ) {
+			if ( symbol.IsReadOnly ) {
+				return true;
+			}
+
+			// For our purposes this is as good as readonly.
+			// Properties in a concisely-defined record like
+			// "record Foo(int X);" will always be init-only rather than
+			// readonly, but there is also regular properties with
+			// "int X { get; init; }".
+			return symbol.SetMethod != null && symbol.SetMethod.IsInitOnly;

```

Comment:
Init-only stuff is actually a bit different:

```cs
using System;

public class Program {
    class Foo { public object X { get; init; } = new object();}
	
    public static void Main() {
        var l = new Foo(){
            X = new Program()
        };
        
        Console.WriteLine(l.X.GetType().FullName);
    }
}
```

Prints "Program". i.e. callers can overwrite the initializers (not just constructors like `readonly`) so we should never narrow our analysis based on initializers.

I'll work on this a bit.